File: go/year2016/sol01.go
package year2016

import (
	"fmt"
	"strings"

	"github.com/dhruvmanila/advent-of-code/go/pkg/set"
	"github.com/dhruvmanila/advent-of-code/go/util"
)

type direction int

const (
	north direction = iota
	east
	south
	west
)

func Sol01(input string) error {
	lines, err := util.ReadLines(input)
	if err != nil {
		return err
	}

	// position is the x and y coordinates representing the current position.
	position := [2]int{0, 0}
	direction := north
	visited := set.New(position)

	// Number of blocks between the initial position and the position which
	// is visited twice. -1 is a sentinel value to indicate that no location
	// has been visited twice yet.
	visitedTwice := -1

	for _, instruction := range strings.Split(lines[0], ", ") {
		switch instruction[0] {
		case 'L':
			direction = util.Mod(direction-1, 4)
		case 'R':
			direction = util.Mod(direction+1, 4)
		default:
			panic("invalid turn: '" + string(instruction[0]) + "'")
		}

		distance := util.MustAtoi(instruction[1:])
		for i := 0; i < distance; i++ {
			switch direction {
			case north:
				position[1]++
			case east:
				position[0]++
			case south:
				position[1]--
			case west:
				position[0]--
			default:
				panic(fmt.Sprintf("invalid direction: %d", direction))
			}
			if visitedTwice == -1 {
				if visited.Contains(position) {
					visitedTwice = util.Abs(position[0]) + util.Abs(position[1])
				} else {
					visited.Add(position)
				}
			}
		}
	}

	blocks := util.Abs(position[0]) + util.Abs(position[1])
	fmt.Printf("1.1: %d\n1.2: %d\n", blocks, visitedTwice)
	return nil
}

File: python/year2016/sol01.py
from enum import Enum
from typing import Iterable

import utils


class Direction(int, Enum):
    NORTH = 0
    EAST = 1
    SOUTH = 2
    WEST = 3


def destination_distance(instructions: Iterable[str]) -> tuple[int, int]:
    x, y = 0, 0
    direction = Direction.NORTH
    visited: set[tuple[int, int]] = {(x, y)}
    visited_twice = -1
    for instruction in instructions:
        direction = (direction + (1 if instruction[0] == "R" else -1)) % 4
        for _ in range(int(instruction[1:])):
            match direction:
                case Direction.NORTH:
                    y += 1
                case Direction.EAST:
                    x += 1
                case Direction.SOUTH:
                    y -= 1
                case Direction.WEST:
                    x -= 1
            if visited_twice == -1:
                if (x, y) in visited:
                    visited_twice = abs(x) + abs(y)
                else:
                    visited.add((x, y))
    return abs(x) + abs(y), visited_twice


if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser()
    parser.add_argument("-t", "--test", action="store_true", help="use the test input")
    args = parser.parse_args()

    data = utils.read(year=2016, day=1, test=args.test)
    blocks, visited_twice = destination_distance(data.split(", "))

    print(f"1.1: {blocks}")
    print(f"1.2: {visited_twice}")
